### DISCUSSION TITLE
AuthenticatorBuilder.subject() does not work

### DISCUSSION THREAD

**Discussion by cc8dea2e-4a29-4394-af6d-af0ccc73647a** [Byron/google-apis-rs#487] [Q&A]
Domain-wide delegation is enabled for the organisation in the Google Admin console, but a cant use AuthenticatorBuilder.subject(), i get an error - `Error(GoogleApiError(MissingToken(AuthError(AuthError { error: UnauthorizedClient, error_description: Some("Client is unauthorized to retrieve access tokens using this method, or client not authorized for any of the scopes requested."), error_uri: None }))))`.
```rust
    let gsa = oauth2::read_service_account_key("path.json").await.?;
    let auth = oauth2::ServiceAccountAuthenticator::builder(gsa)
        .subject("username@domain.name")
        .build()
        .await?;

    let hub = Sheets::new(
        hyper::Client::builder().build(
            hyper_rustls::HttpsConnectorBuilder::new()
                .with_native_roots()
                .https_or_http()
                .enable_http1()
                .build(),
        ),
        auth,
    );

```
I'm sure there are no problems with the access settings, because similar code in Python works.

What am I doing wrong?

**Comment by Byron** [comment]
I think the answer might be this part:

> "[..]  or client not authorized for any of the scopes requested."

The library pre-sets authorization scopes based on the API description, and that can often be wrong due to incorrect heuristics. If you set the scopes by hand, this might work. Otherwise, the code is just one big file and can be inspected to see exactly how it's happening.

**Reply by cc8dea2e-4a29-4394-af6d-af0ccc73647a** [reply]
how i can set scopes by hand?
```rust
let scopes = [
        "https://www.googleapis.com/auth/spreadsheets",
        "https://www.googleapis.com/auth/drive",
        "https://www.googleapis.com/auth/cloud-platform"
];
```

**Reply by Byron** [reply]
Each call can have its scopes set using [this method](https://docs.rs/google-sheets4/latest/google_sheets4/api/struct.SpreadsheetDeveloperMetadataSearchCall.html#method.add_scopes). Otherwise a default-scope kicks in which might not be correct.

**Comment by cc8dea2e-4a29-4394-af6d-af0ccc73647a** [comment]
@Byron Yes, it works! Thank you very much, you helped me a lot!
```rust
    let scopes = [
        "https://www.googleapis.com/auth/spreadsheets",
        "https://www.googleapis.com/auth/drive",
        "https://www.googleapis.com/auth/cloud-platform"
        ];

    let result = hub
        .spreadsheets()
        .values_get(gspread_id, gspread_range)
        .major_dimension("ROWS")
        .value_render_option("UNFORMATTED_VALUE")
        .add_scopes(scopes)
        .doit()
        .await?

```
